### ISSUE TITLE
Batch size较小时训练时loss变成NaN

### ISSUE THREAD

**Issue by eeyrw** [yjh0410/PyTorch_YOLOv2#1]
由于显存小，已经把batch size改成8，对应的lr也减小到5e-4。但是训练过程中网络的LOSS还是发散了。
```
~/PyTorch_YOLOv2$ python train.py -ms --batch_size 8 --cuda --lr 5e-4 --num_workers 4
Setting Arguments.. :  Namespace(batch_size=8, cos=False, cuda=True, dataset='voc', debug=False, eval_epoch=10, gamma=0.1, lr=0.0005, momentum=0.9, multi_scale=True, no_warm_up=False, num_workers=4, resume=None, save_folder='weights/', start_epoch=0, tfboard=False, version='yolov2', weight_decay=0.0005, wp_epoch=1)
----------------------------------------------------------
use cuda
use the multi-scale trick ...
Training model on: VOC0712
The dataset size: 16551
----------------------------------------------------------
Let us train yolov2 on the voc dataset ......
[Epoch 1/200][Iter 0/2068][lr 0.000000][Loss: obj 536.39 || cls 6.45 || bbox 8.00 || iou 0.40 || total 550.84 || size 640 || time: 3.04]
[Epoch 1/200][Iter 10/2068][lr 0.000000][Loss: obj 386.69 || cls 8.87 || bbox 12.47 || iou 0.53 || total 408.04 || size 544 || time: 3.85]
[Epoch 1/200][Iter 20/2068][lr 0.000000][Loss: obj 262.14 || cls 4.32 || bbox 6.91 || iou 0.30 || total 273.37 || size 448 || time: 3.21]
[Epoch 1/200][Iter 30/2068][lr 0.000000][Loss: obj 483.31 || cls 7.25 || bbox 10.54 || iou 0.50 || total 501.09 || size 608 || time: 2.83]
[Epoch 1/200][Iter 40/2068][lr 0.000000][Loss: obj 300.48 || cls 4.83 || bbox 8.46 || iou 0.41 || total 313.77 || size 480 || time: 3.60]
[Epoch 1/200][Iter 50/2068][lr 0.000000][Loss: obj 342.87 || cls 7.32 || bbox 9.85 || iou 0.47 || total 360.04 || size 512 || time: 2.77]
[Epoch 1/200][Iter 60/2068][lr 0.000000][Loss: obj 301.03 || cls 7.69 || bbox 14.52 || iou 0.67 || total 323.25 || size 480 || time: 1.87]
[Epoch 1/200][Iter 70/2068][lr 0.000000][Loss: obj 342.61 || cls 10.57 || bbox 14.07 || iou 0.60 || total 367.25 || size 512 || time: 1.80]
[Epoch 1/200][Iter 80/2068][lr 0.000000][Loss: obj 261.88 || cls 5.19 || bbox 6.74 || iou 0.37 || total 273.80 || size 448 || time: 1.88]
........
[Epoch 1/200][Iter 2000/2068][lr 0.000437][Loss: obj 1.48 || cls 1.80 || bbox 4.42 || iou 0.17 || total 7.70 || size 480 || time: 2.17]
[Epoch 1/200][Iter 2010/2068][lr 0.000446][Loss: obj 3.91 || cls 1.04 || bbox 5.37 || iou 0.16 || total 10.32 || size 576 || time: 1.78]
[Epoch 1/200][Iter 2020/2068][lr 0.000455][Loss: obj 2.00 || cls 2.20 || bbox 3.99 || iou 0.18 || total 8.20 || size 544 || time: 2.45]
[Epoch 1/200][Iter 2030/2068][lr 0.000464][Loss: obj 2.98 || cls 2.55 || bbox 4.34 || iou 0.15 || total 9.88 || size 480 || time: 2.24]
[Epoch 1/200][Iter 2040/2068][lr 0.000473][Loss: obj 3.06 || cls 2.68 || bbox 5.10 || iou 0.16 || total 10.83 || size 384 || time: 1.68]
[Epoch 1/200][Iter 2050/2068][lr 0.000483][Loss: obj 3.37 || cls 3.00 || bbox 7.65 || iou 0.27 || total 14.02 || size 608 || time: 1.47]
[Epoch 1/200][Iter 2060/2068][lr 0.000492][Loss: obj 3.65 || cls 1.58 || bbox 4.97 || iou 0.13 || total 10.20 || size 544 || time: 2.69]
[Epoch 2/200][Iter 0/2068][lr 0.000500][Loss: obj 2.68 || cls 3.46 || bbox 5.60 || iou 0.28 || total 11.73 || size 544 || time: 3.71]
[Epoch 2/200][Iter 10/2068][lr 0.000500][Loss: obj 1.66 || cls 1.89 || bbox 3.92 || iou 0.19 || total 7.47 || size 576 || time: 2.34]
[Epoch 2/200][Iter 20/2068][lr 0.000500][Loss: obj 2.86 || cls 2.93 || bbox 10.16 || iou 0.39 || total 15.95 || size 512 || time: 2.46]
[Epoch 2/200][Iter 30/2068][lr 0.000500][Loss: obj 1.72 || cls 2.61 || bbox 4.61 || iou 0.18 || total 8.94 || size 448 || time: 1.87]
........
[Epoch 2/200][Iter 1270/2068][lr 0.000500][Loss: obj 3.21 || cls 3.00 || bbox 5.22 || iou 0.16 || total 11.44 || size 320 || time: 1.67]
[Epoch 2/200][Iter 1280/2068][lr 0.000500][Loss: obj 3.70 || cls 1.64 || bbox 7.64 || iou 0.21 || total 12.97 || size 448 || time: 1.22]
[Epoch 2/200][Iter 1290/2068][lr 0.000500][Loss: obj 2.81 || cls 2.07 || bbox 4.86 || iou 0.15 || total 9.74 || size 416 || time: 1.54]
[Epoch 2/200][Iter 1300/2068][lr 0.000500][Loss: obj 2.75 || cls 1.92 || bbox 4.79 || iou 0.18 || total 9.46 || size 512 || time: 1.55]
[Epoch 2/200][Iter 1310/2068][lr 0.000500][Loss: obj 3.46 || cls 4.03 || bbox 11.47 || iou 0.43 || total 18.96 || size 320 || time: 1.94]
[Epoch 2/200][Iter 1320/2068][lr 0.000500][Loss: obj 2.33 || cls 1.02 || bbox 4.45 || iou 0.11 || total 7.81 || size 416 || time: 1.29]
[Epoch 2/200][Iter 1330/2068][lr 0.000500][Loss: obj 2.44 || cls 2.34 || bbox 3.43 || iou 0.10 || total 8.21 || size 416 || time: 1.51]
[Epoch 2/200][Iter 1340/2068][lr 0.000500][Loss: obj 2.02 || cls 2.37 || bbox 8.64 || iou 0.48 || total 13.03 || size 352 || time: 1.52]
[Epoch 2/200][Iter 1350/2068][lr 0.000500][Loss: obj 2.97 || cls 0.74 || bbox 4.33 || iou 0.11 || total 8.04 || size 320 || time: 1.34]
[Epoch 2/200][Iter 1360/2068][lr 0.000500][Loss: obj 1.79 || cls 1.54 || bbox 3.09 || iou 0.12 || total 6.42 || size 576 || time: 1.42]
[Epoch 2/200][Iter 1370/2068][lr 0.000500][Loss: obj nan || cls nan || bbox nan || iou nan || total nan || size 608 || time: 2.56]
[Epoch 2/200][Iter 1380/2068][lr 0.000500][Loss: obj nan || cls nan || bbox nan || iou nan || total nan || size 576 || time: 2.74]
[Epoch 2/200][Iter 1390/2068][lr 0.000500][Loss: obj nan || cls nan || bbox nan || iou nan || total nan || size 384 || time: 2.41]
[Epoch 2/200][Iter 1400/2068][lr 0.000500][Loss: obj nan || cls nan || bbox nan || iou nan || total nan || size 448 || time: 1.38]
[Epoch 2/200][Iter 1410/2068][lr 0.000500][Loss: obj nan || cls nan || bbox nan || iou nan || total nan || size 416 || time: 1.62]

```

**Comment by yjh0410** [comment]
@eeyrw 小batch size确实很容易nan，我猜测是我写的loss函数可能有问题。推荐你使用我的另一个项目：
https://github.com/yjh0410/PyTorch_YOLO-Family

不会出现nan问题，训练稳定

**Comment by eeyrw** [comment]
> @eeyrw 小batch size确实很容易nan，我猜测是我写的loss函数可能有问题。推荐你使用我的另一个项目： https://github.com/yjh0410/PyTorch_YOLO-Family
> 
> 不会出现nan问题，训练稳定

在看了。。感觉风格突变。

**Comment by eeyrw** [comment]
终于用lr=2e-4可以安稳的炼完丹，指标供参考：
Evaluating detections
VOC07 metric? Yes
AP for aeroplane = 0.8163
AP for bicycle = 0.8432
AP for bird = 0.7959
AP for boat = 0.7014
AP for bottle = 0.5844
AP for bus = 0.8588
AP for car = 0.8621
AP for cat = 0.8856
AP for chair = 0.6234
AP for cow = 0.8421
AP for diningtable = 0.7208
AP for dog = 0.8589
AP for horse = 0.8767
AP for motorbike = 0.8314
AP for person = 0.8040
AP for pottedplant = 0.5521
AP for sheep = 0.8306
AP for sofa = 0.7205
AP for train = 0.8495
AP for tvmonitor = 0.8029
Mean AP = 0.7830
Mean AP:  0.7830301773517778
Saving state, epoch: 200

**Comment by yjh0410** [comment]
@eeyrw (๑•̀ㅂ•́)و✧！！这个是这个YOLOv2项目训练出来的吗？batchsize=8，lr=2e-4？

**Comment by eeyrw** [comment]
> @eeyrw (๑•̀ㅂ•́)و✧！！这个是这个YOLOv2项目训练出来的吗？batchsize=8，lr=2e-4？

确信。
`python train.py -ms --batch_size 8 --cuda --lr 2e-4 --num_workers 4`

**Comment by yjh0410** [comment]
@eeyrw 辛苦啦！解决了我的心头大困惑~十分感谢您的贡献！
